core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -139,8 +163,17 @@ public void request(int numMessages) {
     if (isClosed()) {
       return;
     }
-    pendingDeliveries += numMessages;
-    deliver();
+    // Make the number of requested messages visible to the transport thread.
+    transportCount.addAndGet(numMessages);
+    applicationCount += numMessages;
+    if (!appInRequest) {
+      appInRequest = true;
```

Comment:
It's intended to only be mutated in the application thread so it depends if we want to allow for multiple application threads to enter the request method. I'm inclined to disallow that in the contract